src/content/symbol-store-db.js

```diff
@@ -1,134 +1,203 @@
-      return this._setupDBPromise.then(() => this.getSymbolsForAddressesInLib(requestedAddressesIndices, libKey));
-    }
+  _findRecordsLastUsedBeforeDate(store: IDBObjectStore, beforeDate: Date, callback: (number[]) => void): void {
+    const lastUsedDateIndex = store.index('lastUsedDate');
+    // Get a cursor that walks all records whose lastUsedDate is less than beforeDate.
+    const cursorReq = lastUsedDateIndex.openKeyCursor(
+      window.IDBKeyRange.upperBound(beforeDate, true));
+    // Iterate over all records in this cursor and collect their primary
+    // keys into allFoundKeys.
+    const allFoundKeys = [];
+    cursorReq.onsuccess = () => {
+      const cursor: IDBCursor = (cursorReq: any).result;
+      if (cursor) {
+        allFoundKeys.push(cursor.primaryKey);
+        cursor.continue();
+        return;
+      }
+      callback(allFoundKeys);
+    };
+  }
 
-    return new Promise((resolve, reject) => {
-      const transaction = this._db.transaction('symbol-tables', 'readonly');
-      const store = transaction.objectStore('symbol-tables');
-      const req = store.get(libKey);
-      req.onerror = reject;
-      req.onsuccess = () => {
-        if (!req.result) {
-          reject(new Error('unexpected null result'));
+  _findNLeastRecentlyUsedRecords(store: IDBObjectStore, n: number, callback: (number[]) => void): void {
+    // Get a cursor that walks the records from oldest to newest
+    // lastUsedDate.
+    const lastUsedDateIndex = store.index('lastUsedDate');
+    const cursorReq = lastUsedDateIndex.openKeyCursor();
+    const allFoundKeys = [];
+    cursorReq.onsuccess = () => {
+      const cursor: IDBCursor = (cursorReq: any).result;
+      if (cursor) {
+        allFoundKeys.push(cursor.primaryKey);
+        if (allFoundKeys.length < n) {
+          cursor.continue();
           return;
         }
-        const { index, buffer } = req.result;
-        const decoder = new TextDecoder();
-        resolve(requestedAddressesIndices.map(addrIndex => {
-          const startOffset = index[addrIndex];
-          const endOffset = index[addrIndex + 1];
-          const subarray = buffer.subarray(startOffset, endOffset);
-          return decoder.decode(subarray);
-        }));
-      };
+      }
+      callback(allFoundKeys);
+    };
+  }
+
+  _count(store: IDBObjectStore, callback: (number) => void): void {
+    const countReq = store.count();
+    countReq.onsuccess = () => callback((countReq: any).result);
+  }
+
+  _deleteLeastRecentlyUsedUntilCountIsNoMoreThanN(store: IDBObjectStore, n: number, callback: () => void): void {
+    this._count(store, symbolTableCount => {
+      if (symbolTableCount > n) {
+        // We'll need to remove at least one symbol table.
+        const needToRemoveCount = symbolTableCount - n;
+        this._findNLeastRecentlyUsedRecords(store, needToRemoveCount, primaryKeysToDelete => {
+          this._deleteRecords(store, primaryKeysToDelete, callback);
+        });
```

Comment:
same comment about calling `cursor.delete()` directly in the loop instead of doing a 2-step operation.